File: libraries/CSharp/Services/GroupService.cs
using System.Net.Http.Json;
using GPMLoginGlobal.Models;

namespace GPMLoginGlobal.Services;

/// <summary>
/// Provides methods for managing profile groups via the GPMLogin Global API.
/// </summary>
public sealed class GroupService(HttpClient http)
{
    /// <summary>
    /// Retrieves all groups.
    /// </summary>
    /// <param name="cancellationToken">Token to cancel the request.</param>
    /// <returns>A read-only list of <see cref="Group"/> objects.</returns>
    /// <exception cref="InvalidOperationException">Thrown when the API returns <c>success: false</c>.</exception>
    public async Task<IReadOnlyList<Group>> GetAllAsync(CancellationToken cancellationToken = default)
    {
        ApiResponse<IReadOnlyList<Group>>? response =
            await http.GetFromJsonAsync<ApiResponse<IReadOnlyList<Group>>>("groups", cancellationToken);

        EnsureSuccess(response);
        return response!.Data!;
    }

    /// <summary>
    /// Retrieves a single group by its identifier.
    /// </summary>
    /// <param name="id">The group identifier.</param>
    /// <param name="cancellationToken">Token to cancel the request.</param>
    /// <returns>The matching <see cref="Group"/>.</returns>
    /// <exception cref="InvalidOperationException">Thrown when the API returns <c>success: false</c>.</exception>
    public async Task<Group> GetByIdAsync(string id, CancellationToken cancellationToken = default)
    {
        ApiResponse<Group>? response =
            await http.GetFromJsonAsync<ApiResponse<Group>>($"groups/{id}", cancellationToken);

        EnsureSuccess(response);
        return response!.Data!;
    }

    /// <summary>
    /// Creates a new group.
    /// </summary>
    /// <param name="request">The group details.</param>
    /// <param name="cancellationToken">Token to cancel the request.</param>
    /// <returns>The newly created <see cref="Group"/>.</returns>
    /// <exception cref="InvalidOperationException">Thrown when the API returns <c>success: false</c>.</exception>
    public async Task<Group> CreateAsync(GroupRequest request, CancellationToken cancellationToken = default)
    {
        HttpResponseMessage httpResponse =
            await http.PostAsJsonAsync("groups/create", request, cancellationToken);
        httpResponse.EnsureSuccessStatusCode();

        ApiResponse<Group>? response =
            await httpResponse.Content.ReadFromJsonAsync<ApiResponse<Group>>(cancellationToken);

        EnsureSuccess(response);
        return response!.Data!;
    }

    /// <summary>
    /// Updates an existing group.
    /// </summary>
    /// <param name="id">The group identifier.</param>
    /// <param name="request">The updated group details.</param>
    /// <param name="cancellationToken">Token to cancel the request.</param>
    /// <returns>The updated <see cref="Group"/>.</returns>
    /// <exception cref="InvalidOperationException">Thrown when the API returns <c>success: false</c>.</exception>
    public async Task<Group> UpdateAsync(string id, GroupRequest request, CancellationToken cancellationToken = default)
    {
        HttpResponseMessage httpResponse =
            await http.PostAsJsonAsync($"groups/update/{id}", request, cancellationToken);
        httpResponse.EnsureSuccessStatusCode();

        ApiResponse<Group>? response =
            await httpResponse.Content.ReadFromJsonAsync<ApiResponse<Group>>(cancellationToken);

        EnsureSuccess(response);
        return response!.Data!;
    }

    /// <summary>
    /// Deletes a group by its identifier.
    /// </summary>
    /// <param name="id">The group identifier.</param>
    /// <param name="cancellationToken">Token to cancel the request.</param>
    /// <exception cref="InvalidOperationException">Thrown when the API returns <c>success: false</c>.</exception>
    public async Task DeleteAsync(string id, CancellationToken cancellationToken = default)
    {
        ApiResponse<object>? response =
            await http.GetFromJsonAsync<ApiResponse<object>>($"groups/delete/{id}", cancellationToken);

        EnsureSuccess(response);
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    private static void EnsureSuccess<T>(ApiResponse<T>? response)
    {
        if (response is null)
            throw new InvalidOperationException("The API returned an empty response.");

        if (!response.Success)
            throw new InvalidOperationException($"GPMLogin API error: {response.Message}");
    }
}

File: libraries/Js/src/services/GroupService.js
/**
 * @file GroupService.js
 * Service for managing profile groups via the GPMLogin Global API.
 */

/**
 * Asserts that `response.success` is truthy, throwing otherwise.
 *
 * @param {object} response - Parsed API response envelope.
 * @throws {Error} When `response.success` is falsy.
 */
function ensureSuccess(response) {
  if (!response.success) {
    throw new Error(`GPMLogin API error: ${response.message}`);
  }
}

/**
 * Provides CRUD operations for profile groups.
 */
export class GroupService {
  /**
   * @param {{ get: Function, post: Function }} http - Shared HTTP helper.
   */
  constructor(http) {
    this._http = http;
  }

  /**
   * Retrieves all profile groups.
   *
   * @returns {Promise<Array<object>>} Array of group objects.
   * @throws {Error} On API failure.
   *
   * @example
   * const groups = await client.groups.getAll();
   */
  async getAll() {
    const response = await this._http.get('groups');
    ensureSuccess(response);
    return response.data;
  }

  /**
   * Retrieves a single group by its identifier.
   *
   * @param {string} id - The group identifier.
   * @returns {Promise<object>} The group object.
   * @throws {Error} On API failure.
   */
  async getById(id) {
    const response = await this._http.get(`groups/${id}`);
    ensureSuccess(response);
    return response.data;
  }

  /**
   * Creates a new profile group.
   *
   * @param {object} params - Group parameters.
   * @param {string} params.name - Display name of the group.
   * @param {number} [params.order] - Optional sort order.
   * @returns {Promise<object>} The newly created group.
   * @throws {Error} On API failure.
   *
   * @example
   * const group = await client.groups.create({ name: 'My Group', order: 1 });
   */
  async create({ name, order }) {
    const body = { name };
    if (order !== undefined) {
      body.order = order;
    }
    const response = await this._http.post('groups/create', body);
    ensureSuccess(response);
    return response.data;
  }

  /**
   * Updates an existing profile group.
   *
   * @param {string} id - The group identifier.
   * @param {object} params - Updated group parameters.
   * @param {string} params.name - New display name.
   * @param {number} [params.order] - New sort order.
   * @returns {Promise<object>} The updated group.
   * @throws {Error} On API failure.
   */
  async update(id, { name, order }) {
    const body = { name };
    if (order !== undefined) {
      body.order = order;
    }
    const response = await this._http.post(`groups/update/${id}`, body);
    ensureSuccess(response);
    return response.data;
  }

  /**
   * Deletes a profile group by its identifier.
   *
   * @param {string} id - The group identifier.
   * @returns {Promise<void>}
   * @throws {Error} On API failure.
   */
  async delete(id) {
    const response = await this._http.get(`groups/delete/${id}`);
    ensureSuccess(response);
  }
}
